### PR TITLE
Feature/unr 3774 detect out of date snapshots

### PR DESCRIPTION
**Pull request by martin-improbable**
#### Description
Added support for snapshot versioning.
Snapshot versions are verified at runtime, on server and client, to ensure they match expected values.

#### Tests
Manual and automated - automated testing only verifies snapshot versioning against expected schema.

#### Documentation
'pull-request-template.md' and 'in code'.

### CODE REVIEW

File: SpatialGDK/Extras/schema/global_state_manager.schema [hunk 1]
```diff
@@ -18,6 +18,11 @@ component DeploymentMap {
     uint32 schema_hash = 4;
 }
 
+component SnapshotVersion {
+    id = 9990;
+    uint32 version_number = 1;
```
Comment (mironec) [inline SpatialGDK/Extras/schema/global_state_manager.schema:23]:
```suggestion
    uint64 version_number = 1;
```
As I think the rest of the code treats it that way?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SnapshotVersion.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//	(a) (optional) if the schema has changed, update SPATIAL_SNAPSHOT_SCHEMA_HASH,
+//			see the test
+//'GIVEN_snapshot_affecting_schema_files_WHEN_hash_of_file_contents_is_generated_THEN_hash_matches_expected_snapshot_version_hash'
+// which will provide the hash on failure
+//  (b) (mandatory) increment SPATIAL_SNAPSHOT_VERSION_INC
+
+constexpr uint32 SPATIAL_SNAPSHOT_SCHEMA_HASH = 958184518;
+constexpr uint32 SPATIAL_SNAPSHOT_VERSION_INC = 1;
+
+constexpr uint64 SPATIAL_SNAPSHOT_VERSION = ((((uint64)SPATIAL_SNAPSHOT_SCHEMA_HASH) << 32) | SPATIAL_SNAPSHOT_VERSION_INC);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SnapshotVersion.h:17]: I think we would want this in SpatialServicesConstants.h instead?
The engine and gdk versions are in their own header files as a special case, to hopefully catch compilation errors earlier, but most of our other GDK-wide constants live in the above file.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SnapshotVersion.h:17]: Or SpatialConstants

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp [hunk 3]
```diff
@@ -1225,4 +1227,135 @@ SCHEMA_GENERATOR_TEST(GIVEN_actor_class_WHEN_generating_schema_THEN_expected_com
 	return true;
 }
 
+SCHEMA_GENERATOR_TEST(
+	GIVEN_snapshot_affecting_schema_files_WHEN_hash_of_file_contents_is_generated_THEN_hash_matches_expected_snapshot_version_hash)
+{
+	SchemaTestFixture Fixture;
+
+	// GIVEN
+	FString GDKSchemaCopyDir = FPaths::Combine(SpatialGDKServicesConstants::SpatialOSDirectory, TEXT("schema/unreal/gdk"));
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp:1236]: I see `"schema/unreal/gdk"` used a few times throughout the codebase, may be worth refactoring into its own constant somewhere

File: SpatialGDK/Extras/schema/global_state_manager.schema (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Extras/schema/global_state_manager.schema:23]: That's a lot of versions

Comment (martin-improbable) [reply SpatialGDK/Extras/schema/global_state_manager.schema:23]: It is version and a schema hash.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp (hunk 3 above)
Comment (martin-improbable) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp:1236]: Good idea, best to tackle in a separate PR, could you add a task for this?

File: SpatialGDK/Extras/schema/global_state_manager.schema [hunk 4]
```diff
@@ -18,6 +18,11 @@ component DeploymentMap {
     uint32 schema_hash = 4;
 }
 
+component SnapshotVersion {
```
Comment (m-samiec) [inline SpatialGDK/Extras/schema/global_state_manager.schema:21]: Please increment `RequireSetup`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -648,6 +649,27 @@ bool USpatialNetDriver::ClientCanSendPlayerSpawnRequests()
 void USpatialNetDriver::OnGSMQuerySuccess()
 {
 	StartupClientDebugString.Empty();
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:652]: Driveby - could you rename OnGSMQuerySuccess to ClientOnGSMQuerySuccess. Helps clarify who we expect this to be called on.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -648,6 +649,27 @@ bool USpatialNetDriver::ClientCanSendPlayerSpawnRequests()
...
+	const uint64 SnapshotVersion = GlobalStateManager->GetSnapshotVersion();
+	if (SPATIAL_SNAPSHOT_VERSION != SnapshotVersion) // Are we running with the same snapshot version?
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error,
+			   TEXT("Your client's snapshot version does not match your deployment's snapshot version. Client version: = '%uu', Server "
+					"version = '%uu'"),
+			   SnapshotVersion, SPATIAL_SNAPSHOT_VERSION);
+
+		if (USpatialGameInstance* GameInstance = GetGameInstance())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:661]: nit - this functionality is dubbed below - worth wrapping in a function. I'm also surprised by the conditional return (in the existing code). Seems we should return regardless.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 7]
```diff
@@ -74,6 +75,32 @@ void UGlobalStateManager::ApplyDeploymentMapData(Schema_ComponentData* Data)
 	SchemaHash = Schema_GetUint32(ComponentObject, SpatialConstants::DEPLOYMENT_MAP_SCHEMA_HASH);
 }
 
+void UGlobalStateManager::ApplySnapshotVersionData(Schema_ComponentData* Data)
+{
+	Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data);
+
+	SnapshotVersion = Schema_GetUint64(ComponentObject, SpatialConstants::SNAPSHOT_VERSION_NUMBER_ID);
+
+	if (NetDriver != nullptr && NetDriver->IsServer())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp:84]: As this gets called on both clients and server, any reason to not check both here?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 8]
```diff
@@ -589,6 +616,11 @@ void UGlobalStateManager::ApplyDeploymentMapDataFromQueryResponse(const Worker_E
 		{
 			ApplyDeploymentMapData(Data.schema_type);
 		}
+
+		if (Data.component_id == SpatialConstants::SNAPSHOT_VERSION_COMPONENT_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp:620]: can't be both - else if ?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp:84]: Oh I guess the query callback will be issues regardless - and we wouldn't want it to do any functionality :hmm:

File: SpatialGDK/Source/SpatialGDK/Public/Schema/StandardLibrary.h [hunk 9]
```diff
@@ -177,6 +177,45 @@ struct Persistence : AbstractMutableComponent
 	}
 };
 
+struct SnapshotVersion : AbstractMutableComponent
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/StandardLibrary.h:180]: StandardLibrary is used for the standard library schema. This is schema definited in the global_state_manager.schema, so should go in it's own file. Could be grouped with other gsm c data definitions, but I'm not sure we have any.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SnapshotVersion.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+// SPATIAL_SNAPSHOT_VERSION is the current version of supported snapshots.
+//
+// If changes are made to the schema or content of snapshots then:
+//	(a) (optional) if the schema has changed, update SPATIAL_SNAPSHOT_SCHEMA_HASH,
+//			see the test
+//'GIVEN_snapshot_affecting_schema_files_WHEN_hash_of_file_contents_is_generated_THEN_hash_matches_expected_snapshot_version_hash'
+// which will provide the hash on failure
+//  (b) (mandatory) increment SPATIAL_SNAPSHOT_VERSION_INC
+
+constexpr uint32 SPATIAL_SNAPSHOT_SCHEMA_HASH = 958184518;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SnapshotVersion.h:14]: Huh, really clever - nice work

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp [hunk 11]
```diff
@@ -1225,4 +1227,135 @@ SCHEMA_GENERATOR_TEST(GIVEN_actor_class_WHEN_generating_schema_THEN_expected_com
...
+	TArray<FString> GDKSchemaFilePaths = { "global_state_manager.schema", "spawner.schema", "virtual_worker_translation.schema" };
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp:1237]: Strings should be wrapped in TEXT() macros

**Review by m-samiec — APPROVED** [review]
Nice work!

File: CHANGELOG.md [hunk 12]
```diff
@@ -20,6 +20,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Enhanced server logging to include load balancing and local worker info on startup.
 - Added 'Persistent' spatial class flag, typically used to override a non persistent base class.
 - Added a button to generate functional test maps from the editor. It can be found under **Window** > **Generate test maps**.
+- Added support for snapshot versioning, which enables 'out of date' snapshots to be flagged on servers and clients.
```
Comment (oblm) [inline CHANGELOG.md:23]: How do I, an UnrealGDK customer with a game project, use this feature?

I'd suggest skimming [this](https://github.com/spatialos/UnrealGDK/blob/master/SpatialGDK/Extras/internal-documentation/how-to-write-good-release-notes.md#guiding-principles) and applying the WIIFM (what's in it for me?) here.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp:1236]: https://improbableio.atlassian.net/browse/UNR-5129

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h [hunk 13]
```diff
@@ -504,6 +507,19 @@ const TArray<Worker_ComponentId> KnownEntityAuthorityComponents = { POSITION_COM
 																	GSM_SHUTDOWN_COMPONENT_ID,	 VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID,
 																	SERVER_WORKER_COMPONENT_ID };
 
+// SPATIAL_SNAPSHOT_VERSION is the current version of supported snapshots.
+//
+// If changes are made to the schema or content of snapshots then:
+//	(a) (optional) if the schema has changed, update SPATIAL_SNAPSHOT_SCHEMA_HASH,
```
[the remark this answers is not in the document]
Comment (martin-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:513]: Snapshots can become invalid for one or more reasons, e.g. changing schema, adding entities, etc. 

This PR only validates schema changes, other validation will be added later.

Therefore, if you change schema you should do parts (a) and (b), however if you add entities to a default snapshot (say) but don't change the schema you should do part (b) only.

[the remark this answers is not in the document]
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:513]: That makes sense, maybe add that to the comment?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp [hunk 14]
```diff
@@ -1225,4 +1227,49 @@ SCHEMA_GENERATOR_TEST(GIVEN_actor_class_WHEN_generating_schema_THEN_expected_com
...
+	// GIVEN
+	FString GDKSchemaCopyDir = FPaths::Combine(SpatialGDKServicesConstants::SpatialOSDirectory, TEXT("schema/unreal/gdk"));
+	TArray<FString> GDKSchemaFilePaths = { TEXT("global_state_manager.schema"), TEXT("spawner.schema"),
+										   TEXT("virtual_worker_translation.schema") };
+
+	// WHEN
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+
+	TArray<FString> FoundSchemaFiles;
+	PlatformFile.FindFilesRecursively(FoundSchemaFiles, *GDKSchemaCopyDir, TEXT(""));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDKEditor/SpatialGDKEditorSchemaGenerator/SpatialGDKEditorSchemaGeneratorTest.cpp:1244]: This isn't used anymore

**Review by m-samiec — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:513]: Additionally I think it would make sense to move the hash generating code here or in a related helper file it feels a bit odd encapsulate that into a test. Maybe it should even be a separate script / helper utility all together but perhaps that is over engineering it.

[the remark this answers is not in the document]
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:513]: Thinking about it its pretty minor at this point, maybe more for future reference I think eventually it would be nice to have an external script for generating hashes that maybe even can be automated as part of the build script?

[the remark this answers is not in the document]
Comment (martin-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:513]: yup definitely, good points, this is only step 1.

**Review by simonsarginson — APPROVED** [review]
LGTM nice work!
